### PR TITLE
Support pretty-print for errors in failures and refactor UnitTest on Combine 

### LINKED ISSUE #119: feature: support Combine
ref: https://developer.apple.com/documentation/combine/publisher/print(_:to:)

My plan is as follows:

```swift
[[1, 2], [3, 4]]
    .publisher
    .prettyPrint("🍎", when: [.request, .value, .completion], format: .multiline, to: nil)
    .sink { _ in }

// =>
// 🍎 request unlimited
// 🍎 receive value:
// [
//     1,
//     2
// ]
// 🍎 receive value:
// [
//     3,
//     4
// ]
// 🍎 receive finished
```

- 1st argument `prefix` and last argument `to` has same functionality of default `print` operator.
- 2nd argument `when` is control when to output.
- 3rd argument `format` is whether to use singleline or multiline as the format.

### PR DESCRIPTION
**Pull request by po-miyasaka**
# As is 
It seems not supporting to `prettyprint` for errors on Combine.
The following is taken when specifing format as .multiline.
Screen Shot 2020-12-09 at 16 37 34

# To be
Developers can see errors just like values.

### with `.multiline`
Screen Shot 2020-12-09 at 16 37 34

### with `.singleline`
Screen Shot 2020-12-09 at 16 39 11

### CODE REVIEW

File: Example/SwiftPrettyPrintExample/Source/ContentViewModel.swift [hunk 1]
```diff
@@ -13,30 +13,34 @@ final class ContentViewModel: ObservableObject {
...
-            Dog(id: DogId(rawValue: "koro"), price: Price(rawValue: 20.0), name: "コロ"),
-        ]
+        let dog1 = Dog(id: DogId(rawValue: "pochi"), price: Price(rawValue: 10.0), name: "ポチ")
+        let dog2 = Dog(id: DogId(rawValue: "koro"), price: Price(rawValue: 20.0), name: "コロ")
 
-        array
+        [dog1]
             .publisher
-            .prettyPrint("🍌", when: [.output, .completion], format: .multiline)
+            .prettyPrint("🐕", when: [.output, .completion], format: .multiline)
```
Comment (po-miyasaka) [inline Example/SwiftPrettyPrintExample/Source/ContentViewModel.swift:21]: This is dog's publisher, so I arranged it to use a dog picture😄

File: Example/SwiftPrettyPrintExample/Source/Dog.swift [hunk 2]
```diff
@@ -14,3 +14,8 @@ struct Dog {
 
 struct DogId { var rawValue: String }
 struct Price { var rawValue: Double }
+
+struct DogsError: Error {
+    let code: Int = 101
+    let message: String = "dogs have run away"
```
Comment (po-miyasaka) [inline Example/SwiftPrettyPrintExample/Source/Dog.swift:20]: It has no means🐕

File: Sources/Public/CombineExtension.swift [hunk 3]
```diff
@@ -63,8 +63,26 @@
...
+                switch completion {
+                case .finished:
+                    _print("receive finished", type: .completion)
+                case let .failure(error):
+                    switch format {
+                    case .singleline:
+                        var s: String = ""
+                        Swift.print("receive failure: ", terminator: "", to: &s)
+                        Pretty.print(error, option: option, to: &s)
+                        _print(s, type: .output, terminator: "")
```
Comment (YusukeHosonuma) [inline Sources/Public/CombineExtension.swift]: `.completion` is correct.

```suggestion
                        _print(s, type: .completion, terminator: "")
```

File: Sources/Public/CombineExtension.swift [hunk 4]
```diff
@@ -63,8 +63,26 @@
...
+
+                    case .multiline:
+                        var s: String = ""
+                        Swift.print("receive failure:", to: &s)
+                        Pretty.prettyPrint(error, option: option, to: &s)
+                        _print(s, type: .output, terminator: "")
```
Comment (YusukeHosonuma) [inline Sources/Public/CombineExtension.swift]:
```suggestion
                        _print(s, type: .completion, terminator: "")
```

File: Sources/Public/CombineExtension.swift [hunk 5]
```diff
@@ -63,8 +63,26 @@
...
                     _print(s, type: .output, terminator: "")
                 }
-            }, receiveCompletion: {
-                _print("receive \($0)", type: .completion)
+            }, receiveCompletion: { completion in
+                switch completion {
+                case .finished:
+                    _print("receive finished", type: .completion)
+                case let .failure(error):
+                    switch format {
```
Comment (YusukeHosonuma) [inline Sources/Public/CombineExtension.swift]: This code pattern is same of L53-65.

I would like to extract local method as follows:

```swift
_prettyPrint(label: "receive failure", type: .completion) // `format` is captured by local method.
```

What do you think?

File: Sources/Public/CombineExtension.swift [hunk 6]
```diff
@@ -63,8 +63,26 @@
                     Pretty.prettyPrint($0, option: option, to: &s)
                     _print(s, type: .output, terminator: "")
                 }
-            }, receiveCompletion: {
-                _print("receive \($0)", type: .completion)
+            }, receiveCompletion: { completion in
```
Comment (YusukeHosonuma) [inline Sources/Public/CombineExtension.swift:69]: Please add unit tests to:
https://github.com/YusukeHosonuma/SwiftPrettyPrint/blob/master/Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift

**Review by YusukeHosonuma — CHANGES_REQUESTED** [review]
This is great improvements!

I made some comments so please check it.

File: Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift [hunk 7]
```diff
@@ -177,6 +226,33 @@ final class CombineExtensionTests: XCTestCase {
...
+            """
+            receive failure:
+            TestError(
+                code: 1,
+                message: "This is the error"
+            )
+            """)
+    }
+    
+    private func subscribeAndWait(_ publisher: AnyPublisher<[Int], TestError>, sendHandler: (() -> Void)? = nil) {
```
Comment (po-miyasaka) [inline Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift:260]: extracted common logics in each test cases.

File: Sources/Public/CombineExtension.swift (hunk 6 above)
Comment (po-miyasaka) [reply Sources/Public/CombineExtension.swift:69]: added!

File: Sources/Public/CombineExtension.swift (hunk 5 above)
Comment (po-miyasaka) [reply Sources/Public/CombineExtension.swift]: extracted it

File: Sources/Public/CombineExtension.swift (hunk 4 above)
Comment (po-miyasaka) [reply Sources/Public/CombineExtension.swift]: fixed it

File: Sources/Public/CombineExtension.swift (hunk 3 above)
Comment (po-miyasaka) [reply Sources/Public/CombineExtension.swift]: fixed it

File: Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift (hunk 7 above)
Comment (YusukeHosonuma) [reply Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift:260]: Great 👍

File: Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift [hunk 8]
```diff
@@ -177,6 +226,33 @@ final class CombineExtensionTests: XCTestCase {
...
+        )
+        
+        assertEqualLines(recorder2.contents.split(separator: "\n")[2...].joined(separator: "\n"),
+            """
+            receive failure:
+            TestError(
+                code: 1,
+                message: "This is the error"
+            )
+            """)
```
Comment (YusukeHosonuma) [inline Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift]: You may also want to use the `do` scope.

```swift
// Success
do {
    let recorder = StringRecorder()
   ...
}

// Failure
do {
    let recorder = StringRecorder()
   ...
}
```

I'll leave it to you.

File: Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift [hunk 9]
```diff
@@ -142,23 +191,23 @@ final class CombineExtensionTests: XCTestCase {
...
-                exp.fulfill()
-            })
-            .sink { value in }
-            .store(in: &cancellables)
-        
-        wait(for: [exp], timeout: 3)
         
+        subscribeAndWait(
+            array.publisher.setFailureType(to: TestError.self).prettyPrint(to: recorder).eraseToAnyPublisher()
+        )
```
Comment (YusukeHosonuma) [inline Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift]: I thought the format was inconsistent.

How about a style that breaks the line just before the `.` for consistent of format?

```suggestion
        subscribeAndWait(
            array
                .publisher
                .setFailureType(to: TestError.self)
                .prettyPrint(to: recorder)
                .eraseToAnyPublisher()
        )
```

File: Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift [hunk 10]
```diff
@@ -100,37 +110,76 @@ final class CombineExtensionTests: XCTestCase {
...
+                  """
+            )
+        ]
+        
+        for test in tests {
+            let subject: PassthroughSubject<[Int], TestError> = .init()
+            let recorder = StringRecorder()
+            subscribeAndWait(
+                subject.prettyPrint(when: test.when,
+                                    format: .singleline,to: recorder).eraseToAnyPublisher(),
```
Comment (YusukeHosonuma) [inline Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift]: e.g.

```suggestion
                subject
                    .prettyPrint(when: test.when, format: .singleline, to: recorder)
                    .eraseToAnyPublisher(),
```

File: Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift (hunk 8 above)
Comment (po-miyasaka) [reply Tests/SwiftPrettyPrintTests/Public/CombineExtensionTests.swift]: Thanks! 
It's nice.
I used it

**Comment by po-miyasaka** [comment]
I addressed all comments, so please review again!

**Review by YusukeHosonuma — APPROVED** [review]
Good job 👍

**Comment by po-miyasaka** [comment]
Thanks for reviews!!!

**Pull request merged**